File: javascript/src/problems1_100/valid-sudoku.js
class Problem {
  isValidSudoku(board) {
    throw new Error(`'Not implemented (${board}).`);
  }
}

class Default extends Problem {
  isValidSudoku(board) {
    let rows = new Set();
    let cols = new Set();
    let sub_board = new Set();
    for (let i = 0; i < 9; i++) {
      rows.clear();
      cols.clear();
      for (let j = 0; j < 9; j++) {
        if (board[i][j] !== '.' && rows.has(board[i][j])) {
          return false;
        }
        if (board[j][i] !== '.' && cols.has(board[j][i])) {
          return false;
        }
        rows.add(board[i][j]);
        cols.add(board[j][i]);
        if (i % 3 !== 0 || j % 3 !== 0) {
          continue;
        }
        sub_board.clear();
        for (let i2 = i; i2 < i + 3; i2++) {
          for (let j2 = j; j2 < j + 3; j2++) {
            if (board[i2][j2] !== '.' && sub_board.has(board[i2][j2])) {
              return false;
            }
          }
        }
      }
    }
    return true;
  }
}

export const SOLUTIONS = [new Default()];

File: python/src/problems1_100/valid_sudoku.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def is_valid_sudoku(self, board: list[list[str]]) -> bool:
        pass


class Default(Problem):
    @override
    def is_valid_sudoku(self, board: list[list[str]]) -> bool:
        rows: set[str] = set()
        cols: set[str] = set()
        sub_board: set[set[str]] = set()
        for i in range(9):
            rows.clear()
            cols.clear()
            for j in range(9):
                if board[i][j] != '.' and board[i][j] in rows:
                    return False
                if board[j][i] != '.' and board[j][i] in cols:
                    return False
                rows.add(board[i][j])
                cols.add(board[j][i])
                if i % 3 != 0 or j % 3 != 0:
                    continue
                sub_board.clear()
                for i2 in range(i, i + 3):
                    for j2 in range(j, j + 3):
                        if board[i2][j2] != '.' and board[i2][j2] in sub_board:
                            return False
        return True


SOLUTIONS: list[Problem] = [Default()]
